### ISSUE TITLE
[Feature Request] Ability to read the collection and folder names in sandbox

### ISSUE THREAD

**Issue by supunt** [a85/POSTMan-Chrome-Extension#6926]
**Preventing copy paste errors when it comes to folders**

Some organizations choose to use a folder per endpoint for endpoint specific testing.
More specific tests sits in an outer folder.

the folder name describes the path (route) the content of the folder (test requests) will be calling. 
Currently there is no way to access the folder/collection name in the sandbox.

**i.e. we have 2 endpoints as follows**

_1. prot://host/api_ver/foo/find/{id}
2. prot://host/api_ver/foo/{bar}_

so we have the following folder structure

- api-version (checks if a certain part of the url matches the api version, if does store in env)
    - foo (checks if a next part of the url is 'foo' relative to what we stored in env)
        - TEST_REQUEST

**Proposed solution**

The ability to access the folder name/ collection name in the pre-exec and test scripts.

**Comment by hbertsch** [comment]
I also would appreciate the folder names to be accessible from tests and pre-request scripts. We needed to export the folder names into the test result file: https://community.getpostman.com/t/get-folder-names-of-a-collection/9516/2 and managed it by using a manual (potentially error prone) workaround.

**Comment by onigor** [comment]
+1 for folder names as variable

**Comment by oehm-smith** [comment]
My use case is different.  

I use the request names to pre-configure some details.  For example 'login user 1' would expand to a `{{userName}} and {{userPassword}}`.  Those could be used to perform the login.  

This works great with `pm.info.requestName`.  I'd like to be able to consolidate activities under a folder.  For example:

```
folder: User 1
    request: Login
    request: Update user login page
    request: ... something also with respect to User 1
folder: User 2
    request: Login
    request: Update user login page
    request: ... something also with respect to User 2
```

**Comment by b0dry** [comment]
Come on guys, please, this is a simple issue, but important for many of us! getParentFolderName should be simple functionality. We are waiting for it already more than two years. It could be any locator: Name or ID, full path, doesn't matter. Give us solution how to know where in the folder we are. We have more than 5000 requests in many collections and folders. Now we need to log the information into a centralised logs. We have created an object that send data to there, but I don't want to go over each request and repeatedly adding one and the same code there. This is really annoying situation. And the fact that we cannot use the standard JS libraries because of the sandbox makes the things really not good.

**Comment by tester-xm** [comment]
I need it so very much!

**Comment by omgkotofey** [comment]
+1

**Comment by supunt** [comment]
Well, we waited, got bored and wrote an entire testing framework using python unit-test with endpoint discovery and simple test auto generation as well.

**Comment by algorun77** [comment]
I would also appreciate this feature. We are using a variable to define our endpoints name (folder Clients ? The endpoint is '/clients', etc) and each time we duplicate a folder to create a new endpoint, we must remember to change that currently hard-coded variable in the duplicated folder. It would really help automation to be able to handle the folder name directly !

**Comment by buzitsl** [comment]
It has been almost two years, and this feature is not implemented?

**Comment by dennisnel** [comment]
+1

**Comment by lflucasferreira** [comment]
+1

**Comment by nossSpb** [comment]
vote up +1

**Comment by JelenaRy** [comment]
+1

**Comment by balmeida-nokia** [comment]
I'd go with an API as:
`pm.request.folder` <- Returns the folder where the request resides. I'd call the object `RequestFolder`
RequestFolder would contain:

```
{
"id": An guid of the folder
"name": The name the user gave to the folder
"authorization": identifier of the authorization used
"parent": A pointer to the parent RequestFolder or null if there's none (do not make the collection available here!)
"collection": A pointer to an object representing the collection where the folder resides.
"hasPrerequestScript": boolean on whether it has a pre-request script
"hasTests": boolean on whether it has a tests script
}
```

**Comment by sanjaykrishnanem** [comment]
+1, would really love to have this feature!!!

**Comment by taipeicruz** [comment]
+1

**Comment by my1795** [comment]
+1 please add this

**Comment by jsilverdev** [comment]
+1

**Comment by half-eaten-apple** [comment]
+1, would really love to have this feature!
My Use case is like this-
I am saving all the variables as Collection variables as it is quite easy to run it through newman(CLI).
I have more than 100 API's in my collection, so my  Collection variables end up more than 300 in number.
Lets assume i use a var -  "userid".
Now these variables can be used many times in many different folders.
    Collection
    |-Folder1
       |--API1-var(userid)
    |-Folder2
      |--API2-var(userid)
  
So i want to name my variable uniquely like <folderName_varname>.
Here if i get an easy way to get folder name , it makes life a lot easier.
Else I have to Hardcode the folder name in Pre-request script and remember to change it every time in Pre-Request script if i change the foldername.
Regards,
half-eaten-apple

**Comment by Erwolo** [comment]
+1

**Comment by kilotone** [comment]
While I do not speak for the entirety of my organization, the implementation of this or as an alternative, a modal interactive prompts is a substantial hurdle. My team at  Disney streaming using postman more extensively if steps were taken to implement.

**Comment by codykhon** [comment]
+1 vote

**Comment by jorge-delagarza** [comment]
+1

**Comment by ryancabanas** [comment]
+1

**Comment by dparadisZentelia** [comment]
+1

**Comment by karukenert** [comment]
+1

**Comment by rabin-io** [comment]
+1

**Comment by KHeavyy** [comment]
+1

**Comment by johnshengkpmg** [comment]
+1 my use case is that I organize request by domain (which is folder). and it need to be referenced in endpoint, for example
https://{{baseURL}}/service/{{tenant}}/{{domain_folder}}/v1
domain_folder= Human_Resources, just one example, and I have 69 domains/folders.

**Comment by bagage** [comment]
+1

**Comment by lawlietl4** [comment]
+1 this would be very helpful for a company that I am working for since we have almost 8k tests that utilize the Postman GUI and since we are integrating with Splunk it would make the value for the folder more accessible at a collection level instead of setting it in a pre-request script

**Comment by tiago-depop** [comment]
+1

**Comment by bounteous-alisWard** [comment]
+1. Our test collection features end-to-end tests, and it would be helpful to be able to output the name of the folderName === test scenario name when a test fails, just like we do with the error message and code.

**Comment by vaibhrez12** [comment]
+1. Basic must feature to use postman as API testing tool.

**Comment by supunt** [comment]
Its been more than 3 years since I requested this. 
We waited 2 years, got bored and wrote an entire testing framework using python unit-test with endpoint discovery and simple test auto generation as well + multiple reporting solutions (logging, files op, elastic, db etc...)

**Comment by bounteous-alisWard** [comment]
Oh! Glad someone followed up on this. I recently discovered the [Postman Pro API collection](https://www.postman.com/postman/workspace/postman-public-workspace/documentation/12959542-c8142d51-e97c-46b6-bd77-52bb66712c9a) that lets you hit postman endpoints against your own collection to pull out this data.

While you can technically use this to READ the folder names in a collection (look under Collections and run All, then find the UID for the collection and run Single), not much else you can do with it except READ. (Tried passing folder `id` in setNextRequest() with no luck, but not surprised since that func() wasn't built to do that). You can also parse out individual request GUIDs, but if you have to string compare to find those anyway, you might as well just use string compares in the setNextRequest() function and skip this lookup.

Requires some work to get at the folder names this way inside whatever test collection you have created. It's more than you probably want to have to do to, say, log the folder's name in a test result, so you know what test scenario that request is part of when it fails, but if you really wanted to, this would be one way. It also requires creating an account-specific API key that 'secret' environment fields don't actually hide from anyone consuming/logging the requests, so there's security/risk to consider if you share your collections or envs a lot.

I do hope they still add support via scripting. They offer string compare for individual requests with `setNextRequest()` (and `console.log(request.name);`, so it seems like support for folders in this and other ways shouldn't be difficult to accomplish. But the pro collection might be a solve for some Pro subscribers who happen on this because all they want to do is get at a list of folders.

**Comment by johnshengkpmg** [comment]
Hi @abeward, 

I like your idea of "log the folder's name"!

Thank you!

**Comment by alekdavisintel** [comment]
So, looks like it won't be done, right?

**Comment by needcaffeine** [comment]
It would greatly benefit my team to be able to read the name of the collection as well as individual requests so we don't have to hardcode things.

**Comment by adambellasps** [comment]
My needs are right in line with this feature request. I'm looking for the UUID of the Folder, accessible from the pre-request script of that folder. This feature request, if satisfied with not just the name but the UUID as well, would benefit me greatly.

**Comment by nathan-beam** [comment]
+1 on this

**Comment by Nyhilo** [comment]
+1 on this. This would greatly simplify our workflow as our schema changes and gets updated.

**Comment by pelonchasva** [comment]
+1

**Comment by montaremco** [comment]
Yes, we would like to use this feature as well.
Hard coding folder names in pre-request scripts results in double the work and in double the chance of errors since folder names can be easily renamed and changing the pre-request script variable values accordingly will be easily forgotten.

**Comment by formatq** [comment]
+1

**Comment by renanfferreira** [comment]
+1

**Comment by pmitko** [comment]
+1

**Comment by hadarnexus** [comment]
+1

**Comment by KaliswaranNachimuthu** [comment]
+1 Desperately need this feature

**Comment by tarakanovalexey** [comment]
+1 need this feature. Suprisingly that issue has no solution for over 2 y now. Microservice architecture means you are making new folders at the moment new service appears. It's really not convinient to add pre script in each new folder. We are using folder name as service name in our requests. Right now postman has no flexible solution for that.

**Comment by gpokvytyte** [comment]
+1 on this, the feature would be useful

**Comment by Bharathraj029** [comment]
Any Update on this feature.

**Comment by otech47** [comment]
please god

**Comment by theodoremca** [comment]
+1 on this, the feature would be useful

**Comment by supunt** [comment]
After 4 years this is still active. 😒.
We created an entire Python FW to discover endpoints and generate basic tests allowing devs to add custom tests in under 1 month.
If this is a deal breaker for ya, I'd recommend that you do something similar

**Comment by aderayendra** [comment]
+1

**Comment by jalbersol** [comment]
Oh come on, what the hell is so hard about the implementation? This would be a really nice feature to bring a bit more automaticity to large collections.

**Comment by ea-soares** [comment]
+1

**Comment by vedkribhu** [comment]
Hello everyone,

We're excited to share our plans to introduce a new API, `pm.execution.location`, to enhance your experience. This API will provide the complete path of a request in an array format. For instance, if you have a request named R1 inside a folder F1 within a collection C1, executing `pm.execution.location` would return ['C1', 'F1', 'R1'].

Additionally, we're introducing another API, `pm.execution.location.current`, which will return the name of the current item. For instance, when executed in a folder pre-request script, `pm.execution.location.current` will provide the current folder's name.

We appreciate your patience and would love to hear your thoughts on these additions. If you have any suggestions or would like to see further improvements, please let us know. Your feedback is crucial to us, and we want to ensure that these enhancements align with your needs before incorporating them into our next release.

**Comment by balmeida-nokia** [comment]
This will help a whole LOT when making pre-request scripts and even global test scripts. This is a good API and I'm thankful for it!

**Comment by algorun77** [comment]
I agree! This is a long-awaited feature that is very welcome.
Can't wait to see that! 👍🏻

**Comment by andrewstet** [comment]
> Hello everyone,
> 
> We're excited to share our plans to introduce a new API, `pm.execution.location`, to enhance your experience. This API will provide the complete path of a request in an array format. For instance, if you have a request named R1 inside a folder F1 within a collection C1, executing `pm.execution.location` would return ['C1', 'F1', 'R1'].
> 
> Additionally, we're introducing another API, `pm.execution.location.current`, which will return the name of the current item. For instance, when executed in a folder pre-request script, `pm.execution.location.current` will provide the current folder's name.
> 
> We appreciate your patience and would love to hear your thoughts on these additions. If you have any suggestions or would like to see further improvements, please let us know. Your feedback is crucial to us, and we want to ensure that these enhancements align with your needs before incorporating them into our next release.

@vedkribhu Any update or ETA on this? So excited to finally have this feature, thank you!

**Comment by vedkribhu** [comment]
@andrewstet this will be out with next release of Postman (v10.21) which is scheduled to happen by Dec 15.

**Comment by andrewstet** [comment]
@vedkribhu excellent news, thank you!

**Comment by andrewstet** [comment]
@vedkribhu So happy this has been released, thank you again! I'm able to use this as I'd hoped for most of my requests, but I've discovered a scenario where I can't. I'm hoping you can either help or direct me to report this as a bug or new feature request. 

Essentially, I was waiting for this new feature so that I could fully automate the URL construction of my requests via pre-request scripts. I have collections with many folders, each with many requests in them. The folder names match the endpoints, which is why this feature was a key piece (and last piece) of automating the URL. Before this feature, I was manually adding the endpoint (same as folder name) to the URL field of each request. With this new feature, I don't have to do that anymore, and I've now added it to the pre-request script (hooray!). This works for all requests that have params, because the params get added to the URL field of the request. However, if I have a simple request with no params added at the request level, then the URL field is empty, because all the other components of the URL get auto-composed in the pre-request script. When the URL field is empty, postman alerts that the URL cannot be empty. It's very possible that postman is not intended to be used this way, but considering my automation works, it should definitely be possible. The only thing blocking me is the fact that the URL field is not allowed to be empty. It would be great to be able to bypass that requirement, or even have postman be aware at the request level that the URL is being composed in a pre-request script, for example. I've attached screenshots of my setup as visual aides, I hope it's clear. Any help or guidance would be much appreciated, thank you for your time!


<img width="1562" alt="request" src="https://github.com/postmanlabs/postman-app-support/assets/53357401/8c3a2c1c-ce13-415f-9afe-b449e18b47c1">
<img width="1239" alt="authorization" src="https://github.com/postmanlabs/postman-app-support/assets/53357401/575d1f58-6478-4aca-bc47-8802a4100b2f">
<img width="1239" alt="variables" src="https://github.com/postmanlabs/postman-app-support/assets/53357401/d4a54af7-2289-4064-a1a2-0d34c5053d73">
<img width="1239" alt="pre_request_script" src="https://github.com/postmanlabs/postman-app-support/assets/53357401/d2501c6b-831f-4403-a686-133d3e6724de">

**Comment by DSchau** [comment]
@andrewstet sorry, I'm a little confused and may just not understand. In your script, couldn't you just detect the blank case and then set the URL to what you'd like it to be in this case?

e.g. without fully understanding what you want the URL to be, something like this in theory would work

```js
let url = [
  pm.collectionVariables.get('base_url'),
  pm.execution.location[1],
  pm.request.url
].join('')

if (url === '') {
  // set it to something non-null
}

pm.request.url = url
```

I don't think it makes sense for us at Postman to allow blank URLs at this point to support this use case, mostly because I think there's probably more that manual scripting will enable you to provide a workaround!

**Comment by andrewstet** [comment]
@DSchau No need to apologize, thank you for the reply! The issue appears to be before the script, at the request level, potentially purely from the GUI. I can't fire off the request using the `Send` button from the request's GUI when the URL is "empty". I hope these two marked-up screenshots help illustrate the issue more clearly, but if not, I'm happy to elaborate as much as needed. Thank you!

<img width="1562" alt="request_fail" src="https://github.com/postmanlabs/postman-app-support/assets/53357401/b2276fbe-a3ce-486b-995d-823480f1e82d">
<img width="1239" alt="request_success" src="https://github.com/postmanlabs/postman-app-support/assets/53357401/883d329c-e9c4-4167-aa70-b30d0d039bea">

**Comment by DSchau** [comment]
Ah now I get it. You entirely generate the request URL from the script, so Postman is unaware that the request _will_ actually succeed.

I suppose here I'd have one more question. Why not just include the base_url variable for all the requests in the collection? (and then thus remove it from the script) It also makes it easier to understand what's happening for each request as a user of the collection!

**Comment by andrewstet** [comment]
@DSchau Yes, exactly! To your suggestion, if you mean adding the `base_url` as a collection variable, I'm doing that already and using it in the pre-request script to compose the URLs of each request in the collection. Though if you mean adding the base URL to the URL field of each request, I could certainly do that, but it defeats my goal of fully automating the URLs. Perhaps that's a fool's errand haha. Before this new feature of being able to access the folder name in the pre-request script, the folder name (identical to the the endpoint), was the only piece I couldn't automate, so each request's folder name was added to the URL field of each request. If I add the base URL instead, I'm just replacing one for the other, and not getting closer to my goal.


> Postman is unaware that the request will actually succeed

And to this point, this is why I would love the ability to bypass that empty URL field check! Postman could instead say, "hey, your URL is empty and the request will likely fail, proceed anyway?", with an option to always try anyway and not ask again, for example.

**Comment by DSchau** [comment]
@andrewstet ha, yeah I should've included my own screenshots! Here's what I mean.

<img width="438" alt="Screenshot 2023-12-18 at 13 18 41" src="https://github.com/postmanlabs/postman-app-support/assets/3924690/698a857c-3d95-4e9f-a012-d5525aad75f7">

I would've thought the utility of this feature is precisely in getting the path to the collection folder names (which can be super useful to derive a _piece_ of the URL state) and I think we'd generally recommend to users to persist at least some portion of the URL in the URL bar so that it's clearer to the end user of the collection what will happen _before_ running the pre-request script.

So in your scenario, I think I would recommend the following here:

1. "Store" baseURL variable in the request, and remove that portion from the pre-request script
2. Continue to use the pm.execution.location folder to derive the _rest_ of the URL

I know it may seem a bit mundane to store something in the URL state that never changes, but I do think it helps simplify and clarify to the end user here (and nicely also, does not require changes to Postman upstream to build in support _for this_ for the empty URL state). As a specific example, it can be quite nice to tap into other features of Postman -- like if you change an environment, the user can "click" the baseURL in the URL bar and know what the resolved value will be -- instead of deriving it entirely via a script.

(As an example, here's a recent example of how I did this with a sample app, swapping HOST for baseURL. I can hover over and get a better sense of what URL I'll be sending to without needing to click around and get a little confused)

<img width="324" alt="Screenshot 2023-12-18 at 13 30 27" src="https://github.com/postmanlabs/postman-app-support/assets/3924690/b302b43c-93da-4185-bb85-a1edcc2be1be">

**Comment by andrewstet** [comment]
@DSchau Fair enough, makes sense to me, thank you!

**Comment by giridharvc7** [comment]
You can now access the request path in the sandbox with the following: 
* `pm.execution.location` returns the full path of -> Collection,folder,request
* `pm.execution.location.current` returns the current location of execution. 

This is available on the december release of the app, v10.21

**Comment by chandrudev1996** [comment]
Finally, I found a solution for this.
```
/**
 * Folder
 */
const location = pm.execution.location;
const filterLocation = location.filter((f, i) => i && f !== "root" && f !== "index" );
const folder = filterLocation.join("/");
pm.variables.set("folder_name", folder);
```
